cleverhans/attacks.py

```diff
@@ -749,6 +749,91 @@ def parse_params(self, y=None, y_target=None, nb_classes=None,
         self.clip_max = clip_max
 
 
+class DeepFool(Attack):
+
+    """
+    DeepFool is an untargeted & iterative attack which aims at finding the
+    minimum adversarial perturbations in deep networks. The implementation
+    here is w.r.t. the L2 norm.
+    Paper link: "https://arxiv.org/pdf/1511.04599.pdf"
+    """
+
+    def __init__(self, model, back='tf', sess=None):
+        """
+        Create a DeepFool instance.
+        """
+        super(DeepFool, self).__init__(model, back, sess)
+
+        if self.back == 'th':
+            raise NotImplementedError('Theano version not implemented.')
+
+        import tensorflow as tf
+        self.structural_kwargs = ['over_shoot', 'max_iter',
+                                  'clip_max', 'clip_min',
+                                  'nb_candidate', 'nb_classes']
+
+        if not isinstance(self.model, Model):
+            self.model = CallableModelWrapper(self.model, 'logits')
+
+    def generate(self, x, **kwargs):
+        """
+        Generate symbolic graph for adversarial examples and return.
+        :param x: The model's symbolic inputs.
+        :param nb_candidate: The number of classes to test against, i.e.,
+                            deepfool only consider nb_candidate classes when
+                            attacking (thus accelerate speed)
```

Comment:
How are these classes chosen? Are those the first nb_candidate classes that are different from the correct class?

Reply:
These classes are selected based on the prediction confidence, e.g., the classes with confidence among the top nb_candidate are chosen. The first nb_candidate classes is the same as the correct class if the prediction is correct. The parameter nb_candidate is named as parameter NUM_LABELS in the original DeepFool implementation (DeepFool/MATLAB/adversarial_perturbation.m)